### PR TITLE
Fix `The variable 'irrevent' is being used without being initialized` error in Debug build from MSVS

### PR DESCRIPTION
**Pull request by Mahoyomu**
Add compact, short information about your PR for easier understanding:

- Goal of the PR
- Fix https://github.com/minetest/minetest/issues/14779
- How does the PR work?
- By move the declaration into the while loop, like sfan5 said.
- Does it resolve any reported issue?
- https://github.com/minetest/minetest/issues/14779
- Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?
- IMHO yes.
- If not a bug fix, why is this PR needed? What usecases does it solve?
- this is a bug fix, without this fixed, the MSVS debug build will not be able to run, can only use Release build.

## To do
see if members give me advise about other implementation of this functionality or code logic, and help of commiting about code format problems, especially the latter if there's code format problems in any reviewers's opinion. (I'm probably capable of doing the former myself)

This PR is Ready for Review.

## How to test
See the test video, or see changes in src.

### CODE REVIEW

**Comment by appgurueu** [comment]
> By move the declaration into the while loop, like sfan5 said.

I don't think this is what fixes the bug. Note that sfan5's comment was about preventing future iterations of another bug (which is already fixed), so "fixing" that would be a code quality improvement, not a bug fix. It is also somewhat orthogonal to fixing the bug. (Though I imagine a proper refactor from a bare `union` to a `std::variant` might end up accidentally fixing it.)

What this PR does (and how it probably "fixes" the issue) is initialize `irrevent` (the `{}` does [aggregate initialization](https://en.cppreference.com/w/cpp/language/aggregate_initialization)).

We have to consider two cases: Either this is a "true positive", and then we should definitely *not* mask it via a "fix" like this one, replacing garbage data with other garbage data (with unknown implications) to make a (correctly) failing run-time check go away.

Or this is a "false positive" and a workaround like this would be appropriate, but that would require proper research to actually confirm that this is in fact a false positive.

File: irr/src/CIrrDeviceSDL.cpp [hunk 1]
```diff
@@ -700,12 +700,11 @@ bool CIrrDeviceSDL::run()
 {
 	os::Timer::tick();
 
-	SEvent irrevent;
 	SDL_Event SDL_event;
 
 	while (!Close && wrap_PollEvent(&SDL_event)) {
+		SEvent irrevent{};
 		// os::Printer::log("event: ", core::stringc((int)SDL_event.type).c_str(),   ELL_INFORMATION);	// just for debugging
-
```
Comment (JosiahWI) [inline irr/src/CIrrDeviceSDL.cpp:708]: It's a good idea not to make changes unrelated to the commit. I don't understand why this line was removed.

**Comment by sfan5** [comment]
> Or this is a "false positive" and a workaround like this would be appropriate, but that would require proper research to actually confirm that this is in fact a false positive.

I think that's unnecessary. We're either fixing a real bug or we're not and merely making MSVC happy.

File: irr/src/CIrrDeviceSDL.cpp [hunk 2]
```diff
@@ -700,12 +700,11 @@ bool CIrrDeviceSDL::run()
 {
 	os::Timer::tick();
 
-	SEvent irrevent;
 	SDL_Event SDL_event;
 
 	while (!Close && wrap_PollEvent(&SDL_event)) {
+		SEvent irrevent{};
```
Comment (sfan5) [inline irr/src/CIrrDeviceSDL.cpp:706]: A pitfall here is that aggregate-initialization only initializes the first member of an `union`.
what we have to do here is `memset(&irrevent, 0, sizeof(irrevent));` (and then you can also move the declaration back)

**Comment by appgurueu** [comment]
> I think that's unnecessary. We're either fixing a real bug or we're not and merely making MSVC happy.

As I said, if we're "fixing" a real bug, I don't have faith in this PR that it fixes the bug rather than just masking it.

I will gladly accept this PR if someone can explain to me where and why the supposed uninitialized read is happening and why a zero-initialized union is the correct value to be used there.

Conversely, I also don't have faith in the fact that this is a false positive, because if so I would assume that we aren't the only ones affected, but I haven't yet found evidence of this run-time check having false positives (I also haven't looked very hard tbh).

In other words: The current situation is unclear to me, hence I'd be hesitant with merging such a change as a supposed "fix".

At the very least I would advocate for a `// FIXME` or `// HACK` annotation explaining the purpose of this.

File: irr/src/CIrrDeviceSDL.cpp [hunk 3]
```diff
@@ -705,6 +705,7 @@ bool CIrrDeviceSDL::run()
 
 	while (!Close && wrap_PollEvent(&SDL_event)) {
 		// os::Printer::log("event: ", core::stringc((int)SDL_event.type).c_str(),   ELL_INFORMATION);	// just for debugging
+		memset(&irrevent, 0, sizeof(irrevent));
```
Comment (JosiahWI) [inline irr/src/CIrrDeviceSDL.cpp:708]: `std::memset` requires the `<cstring>` header.

**Review by JosiahWI — CHANGES_REQUESTED** [review]

File: irr/src/CIrrDeviceSDL.cpp (hunk 3 above)
Comment (SmallJoker) [reply irr/src/CIrrDeviceSDL.cpp:708]: `irrString` includes `string.h`, thus LGTM.

**Review by SmallJoker — APPROVED** [review]
LGTM, but did not test on MSVC.
Test passed: https://irc.minetest.net/minetest-dev/2024-07-20#i_6186470

File: irr/src/CIrrDeviceSDL.cpp (hunk 3 above)
Comment (appgurueu) [reply irr/src/CIrrDeviceSDL.cpp:708]: I don't think it is a good idea to rely on transitive includes.

**Comment by appgurueu** [comment]
MSVC is short for Microsoft Visual C++. This is what Microsoft Visual Studio (I assume this is what you mean by MSVS) uses.

**Comment by JosiahWI** [comment]
I built the Debug configuration on MSVC and confirmed that there is no error with this patch.
